fedimint-client/src/lib.rs

```diff
@@ -623,29 +623,35 @@ impl Drop for ClientArc {
                 }
 
                 let db = inner.db.clone();
-                let federation_id = inner.federation_id();
 
+                tokio::task::block_in_place(|| {
+                    tokio::runtime::Handle::current().block_on(async {
+                        // futures::executor::block_on(async {
+                        debug!(target: LOG_CLIENT, "Waiting for client task group to shut down");
+                        if let Err(err) = inner
+                            .task_group
+                            .clone()
+                            .shutdown_join_all(Some(Duration::from_secs(30)))
+                            .await
+                        {
+                            warn!(target: LOG_CLIENT, %err, "Error waiting for client task group to shut down");
+                        }
+                    });
+                });
+
+                let client_strong_count = Arc::strong_count(&inner);
+                debug!(target: LOG_CLIENT, "Dropping last handle to Client");
+                // We are sure that no background tasks are running in the client anymore, so we
+                // can drop the (usually) last inner reference.
                 drop(inner);
 
-                // wait until `self.inner.db` is the only strong reference
-                for attempt in 0u64.. {
-                    let strong_count = db.strong_count();
-                    if strong_count <= 1 {
-                        break;
-                    }
-                    tokio::task::block_in_place(|| {
-                        futures::executor::block_on(async {
-                            // we want to retry fast, give feedback, but not spam
-                            if attempt % 100 == 0 {
-                                info!(
-                                    %federation_id,
-                                    strong_count,
-                                    "Waiting for client database to stop being used"
-                                );
-                            }
-                            fedimint_core::task::sleep(Duration::from_millis(10)).await;
-                        });
-                    });
+                if client_strong_count != 1 {
+                    debug!(target: LOG_CLIENT, count = client_strong_count - 1, LOG_CLIENT, "External Client references remaining after last handle dropped");
```

Comment:
Should this be a warning?

Reply:
AFAIU, at least currently it seems possible to get ahold of `Arc<Client>` that would not be a `ClientHandle` from the user code and maybe there's a legitimate reason to do it. E.g. not everyone will need to be restarting the `Client`.

It's even more the case with the `Database` references, which will generally always be possible to hold longer than the `Client` itself.

So for now I just made these debug diagnostics.